# EDAP01 - Assignment Search: Solving a Maze Using Different Search Algorithms

In this assignment, you are supposed to code the following search algorithms to solve various given mazes:

    - Breadth-first search (BFS)
    - Depth-first search (DFS)
    - Uniform-cost search (UCS)
    - Greedy best-first search (GBFS)
    - A* search

Please follow the basic instructions below to complete this notebook. For each search algorithm, report the following findings:

    - The solution and its path cost
    - The total number of nodes expanded
    - The length of the frontier
    - The total memory size to store all nodes
    - The depth size 

Submit the notebook file with all outputs. Check that your submitted file is readable and contains all saved images. Document your code. Use comments in the code to explain how your implementation works and detail your design choices.


This assignment folder (**EDAP01-LabAssignmentSearch.zip**) contains the following content:
 
* `Assignment_Search.ipynb`: This is the current <a href="https://anaconda.org/anaconda/jupyter">Jupyter</a> notebook file that you are reading. It comes with instructions. You need to edit and submit this file.
* `images`: This directory contains the output images.
* `mazes`: This directory contains predefined mazes.
* `searchalgorithms`: This directory contains the Python scripts that you need to code in order to solve different mazes. Each search algorithm has its own Python file (e.g. bfs.py), which is instantiated from the base class **SearchAlgorithmBase**, as defined in base.py. Your algorithms are expected to use the properties and methods of **SearchAlgorithmBase**. You need to complete and submit these Python files in this folder.
* `src/maze.py`: This file contains the maze class with all the required methods to generate a maze. **You should not make any changes to this file.**   
* `main.py`: This is the main file for running your search algorithms on the given maze. **No need to make any changes to this file.** 
* `requirements.txt`: This file has the list of all required libs to be installed.

## Step 1 - Installing Python and  Jupyter Notebook

To work on assignments, you can use one of these environments listed below:

* Use the online service <a href="https://colab.research.google.com/">Google CoLab</a>. No additional installations are necessary. Please note that our animations may not work on Colab due to some incompatibility issues.
* Install <a href="https://code.visualstudio.com/download">Visual Studio Code</a>. It will prompt you to install all the necessary software when you open a notebook file.
* Install Python, <a href="https://jupyter.org/">JupyterLab</a> and all needed packages yourself using pip or conda. Alternatively, you can use <a href="https://anaconda.org/">Anaconda</a>   for easier installation.

## Step 2 - Learning Python and Jupyer Notebook
If you are not familiar with Python, then you should go through the provided Python tutorial to learn the basics of Python and its packages such as numpy. Some code examples that help with the assignments are already available in this assignment.

Many online tutorials, such as the <a href="https://www.dataquest.io/blog/jupyter-notebook-tutorial/">Jupyer notebook tutorial</a>, cover how to use Jupyter notebooks.  

## Step 3 - Installation of the Required Libs

Once you have set up the Python environment, you need to import [numpy] and [pygame](https://www.pygame.org/news). To install these additional libs, run the following command:

In [ ]:
pip install -r requirements.txt

Note: you may need to restart the kernel to use updated packages.


## Step 4 - Sanity check

To make sure that you have the correct setup, lets run the following command below. It will run the search algoritm **example** defined in ./searchalgorithms/example.py on the fifth maze (i.e., maze_id=5). If everything works correctly, a new window will pop up showing the maze with the start and end points indicated in blue and red, respectively. Since this example code is incomplete, no path will be computed. Therefore, the code will output a message indicating that there are zero expanded nodes and cost. To close the new window, just press the **Esc** button. Note that the final version of the map with the found path will be automatically saved as an image in the ./images folder. You should then see the following image: ./images/maze_5_algorithm_example.png. 

In [7]:
!python main.py --maze_id=5 --search_algorithm_name="example"

^C


There are a total of five different mazes, each of which is shown below.

<table style="width:100%; border:none">
  <tr>
    <td style="text-align:center; border:none">
      <img src="./images/maze1.png" width="200" alt="maze_1"  align="center">
      <br>
      <b>1</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/maze2.png" width="200" alt="maze_2"  align="center">
      <br>
      <b>2</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/maze3.png" width="200" alt="maze_3"  align="center">
      <br>
      <b>3</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/maze4.png" width="200" alt="maze_4"  align="center">
      <br>
      <b>4</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/maze5.png" width="200" alt="maze_5"  align="center">
      <br>
      <b>5</b>
    </td>
  </tr>
</table>

Once you successfully complete the required search algorithms, you should see such animations below. These animations are for the fifth maze, with all visited nodes shown in blue, frontiers in red and the final shortest path in green. All obstacles such as walls are depicted in dark brown.
 
 <table style="width:100%; border:none">
  <tr> <img src="./images/legend.png" width="600"  align="left"> </tr>
  <tr>
    <td style="text-align:center; border:none">
      <img src="./images/5bfs.gif" width="200" alt="BFS"  align="center">
      <br>
      <b>BFS</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/5dfs.gif" width="200" alt="DFS"  align="center">
      <br>
      <b>DFS</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/5ucs.gif" width="200" alt="UCS"  align="center">
      <br>
      <b>UCS</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/5gbfs.gif" width="200" alt="GBS"  align="center">
      <br>
      <b>GBFS</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/5astar.gif" width="200" alt="A*"  align="center">
      <br>
      <b>A*</b>
    </td>
  </tr>
</table>

## Step 5 - Tasks

Please follow instructions below and complete each task. Submit your source codes (Python files under the /searchalgorithms folder) together with this notebook with your findings (e.g., tables fillout out below)!

### Task 1: Breadth-First Search 

Follow the pseudocode in the textbook/slides and implement the **Breadth-First Search (BFS)** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/bfs.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your BFS algorithm in this specific file. 
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.


#### Pseudocode for BFS

Function BFS(problem):
    fringe = Queue{Node(problem.initial_state)}
    while (true):
        if (fringe.empty) return fail
        node = fringe.pop() #FIFO
        if (node.state is goal) return solution
        else fringe.add(Expand(node,problem))

In [ ]:
!python main.py --maze_id=1 --search_algorithm_name="bfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.11.4)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
bfs is loaded
------------------------
Found Path with 63 cost
Expanded 346 nodes
Max Frontier size 14
Max node memory size 348
Max depth 63
------------------------


In [ ]:
!python main.py --maze_id=2 --search_algorithm_name="bfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.11.4)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
bfs is loaded
------------------------
Found Path with 105 cost
Expanded 372 nodes
Max Frontier size 9
Max node memory size 375
Max depth 105
------------------------


In [ ]:
!python main.py --maze_id=3 --search_algorithm_name="bfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.11.4)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
bfs is loaded
------------------------
Found Path with 37 cost
Expanded 143 nodes
Max Frontier size 9
Max node memory size 151
Max depth 37
------------------------


In [ ]:
!python main.py --maze_id=4 --search_algorithm_name="bfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.11.4)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze4.txt is loaded
bfs is loaded
------------------------
Found Path with 59 cost
Expanded 695 nodes
Max Frontier size 28
Max node memory size 701
Max depth 59
------------------------


__BFS__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         |    63     |        346     |        63      |          348  |            14     |
| 2         |    105    |        372     |      105       |      375      |         9         |
| 3         |    37     |        143     |     37         |      151      |        9          |
| 4         |    59     |        695     |     59         |      701      |        28         | 


### Task 2: Depth-First Search 

Follow the pseudocode in the textbook/slides and implement the **Depth-First Search (DFS)** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/dfs.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your DFS algorithm in this specific file. 
* Note that you can also use the BFS code here. In this case, however, you need to switch from the **Queue** data structure to the **Stack** data structure. 
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.

In [ ]:
!python main.py --maze_id=1 --search_algorithm_name="dfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
dfs is loaded
------------------------
Found Path with 155 cost
Expanded 283 nodes
Max Frontier size 20
Max node memory size 301
Max depth 154
------------------------


In [ ]:
!python main.py --maze_id=2 --search_algorithm_name="dfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
dfs is loaded
------------------------
Found Path with 161 cost
Expanded 273 nodes
Max Frontier size 26
Max node memory size 298
Max depth 160
------------------------


In [5]:
!python main.py --maze_id=3 --search_algorithm_name="dfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
dfs is loaded
------------------------
Found Path with 37 cost
Expanded 406 nodes
Max Frontier size 36
Max node memory size 410
Max depth 142
------------------------


In [8]:
!python main.py --maze_id=4 --search_algorithm_name="dfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze4.txt is loaded
dfs is loaded
------------------------
Found Path with 217 cost
Expanded 505 nodes
Max Frontier size 139
Max node memory size 616
Max depth 246
------------------------


__DFS__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         | 155       |        283     |     154        |      301      |        20         |
| 2         | 161       |        273     |     160        |      298      |        26         |
| 3         | 37        |        406     |     142        |      410      |        36         |
| 4         | 217       |        505     |     246        |      618      |       139         | 


### Task 3: Uniform-Cost Search 

Follow the pseudocode in the textbook/slides and implement the **Uniform-Cost Search (UCS)** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/ucs.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your UCS algorithm in this specific file. 
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.

In [11]:
!python main.py --maze_id=1 --search_algorithm_name="ucs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
ucs is loaded
------------------------
Found Path with 63 cost
Expanded 346 nodes
Max Frontier size 14
Max node memory size 348
Max depth 63
------------------------


In [12]:
!python main.py --maze_id=2 --search_algorithm_name="ucs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
ucs is loaded
------------------------
Found Path with 105 cost
Expanded 372 nodes
Max Frontier size 9
Max node memory size 375
Max depth 105
------------------------


In [13]:
!python main.py --maze_id=3 --search_algorithm_name="ucs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
ucs is loaded
------------------------
Found Path with 37 cost
Expanded 143 nodes
Max Frontier size 9
Max node memory size 151
Max depth 37
------------------------


In [14]:
!python main.py --maze_id=4 --search_algorithm_name="ucs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze4.txt is loaded
ucs is loaded
------------------------
Found Path with 59 cost
Expanded 695 nodes
Max Frontier size 28
Max node memory size 701
Max depth 59
------------------------


__UCS__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         |  63       |     346        |        63      |    348        |        63         |
| 2         |  105      |     372        |       105      |    375        |        9          |
| 3         |  37       |     143        |        37      |    151        |        9          |
| 4         |  59       |     695        |        59      |    701        |        28         | 


### Task 4: Greedy Best-First Search 

Follow the pseudocode in the textbook/slides and implement the **Greedy Best-First Search (GBFS)** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/gbfs.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your GBFS algorithm in this specific file. Use the **Manhattan distance** as the heuristic function.
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.

In [24]:
!python main.py --maze_id=1 --search_algorithm_name="gbfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
gbfs is loaded
------------------------
Found Path with 67 cost
Expanded 89 nodes
Max Frontier size 16
Max node memory size 103
Max depth 66
------------------------


In [25]:
!python main.py --maze_id=2 --search_algorithm_name="gbfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
gbfs is loaded
------------------------
Found Path with 119 cost
Expanded 164 nodes
Max Frontier size 19
Max node memory size 182
Max depth 118
------------------------


In [26]:
!python main.py --maze_id=3 --search_algorithm_name="gbfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
gbfs is loaded
------------------------
Found Path with 43 cost
Expanded 44 nodes
Max Frontier size 10
Max node memory size 53
Max depth 42
------------------------


In [27]:
!python main.py --maze_id=4 --search_algorithm_name="gbfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze4.txt is loaded
gbfs is loaded
------------------------
Found Path with 59 cost
Expanded 151 nodes
Max Frontier size 45
Max node memory size 195
Max depth 58
------------------------


__GBFS__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         |   67      |      89        |      66        |      103      |       16          |
| 2         |   119     |      164       |      118       |      182      |       19          |
| 3         |    43     |      44        |      42        |      53       |       10          |
| 4         |    59     |      151       |      58        |      195      |       45          | 


### Task 5: A* Search 

Follow the pseudocode in the textbook/slides and implement the **A*** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/astar.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your A* algorithm in this specific file. Use the **Manhattan distance** as the heuristic function.
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.

In [29]:
!python main.py --maze_id=1 --search_algorithm_name="astar"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
astar is loaded
------------------------
Found Path with 63 cost
Expanded 181 nodes
Max Frontier size 16
Max node memory size 186
Max depth 62
------------------------


In [30]:
!python main.py --maze_id=2 --search_algorithm_name="astar"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
astar is loaded
------------------------
Found Path with 105 cost
Expanded 279 nodes
Max Frontier size 9
Max node memory size 284
Max depth 104
------------------------


In [31]:
!python main.py --maze_id=3 --search_algorithm_name="astar"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
astar is loaded
------------------------
Found Path with 37 cost
Expanded 86 nodes
Max Frontier size 10
Max node memory size 95
Max depth 36
------------------------


In [32]:
!python main.py --maze_id=4 --search_algorithm_name="astar"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze4.txt is loaded
astar is loaded
------------------------
Found Path with 59 cost
Expanded 524 nodes
Max Frontier size 27
Max node memory size 539
Max depth 58
------------------------


__A*__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         |     63    |       181      |       62       |      186      |       16          |
| 2         |    105    |       279      |      104       |      284      |       9           |
| 3         |     37    |       86       |       36       |       95      |       10          |
| 4         |     59    |       524      |       58       |      539      |       27          | 


### Task 6: Comparison and discussion

Run all your implemented search algorithms on maze 5 (i.e. maze_id=5) and complete the following table with your findings. Discuss your reported scores in the table. For example, explain which method returns the lowest path cost and why. Likewise, explain which method is more memory friendly and why that is the case. If you have any more discussion points, please do share them. 

 

In [34]:
!python main.py --maze_id=5 --search_algorithm_name="bfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze5.txt is loaded
bfs is loaded
------------------------
Found Path with 51 cost
Expanded 576 nodes
Max Frontier size 22
Max node memory size 581
Max depth 51
------------------------


In [35]:
!python main.py --maze_id=5 --search_algorithm_name="dfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze5.txt is loaded
dfs is loaded
------------------------
Found Path with 155 cost
Expanded 353 nodes
Max Frontier size 149
Max node memory size 465
Max depth 198
------------------------


In [36]:
!python main.py --maze_id=5 --search_algorithm_name="ucs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze5.txt is loaded
ucs is loaded
------------------------
Found Path with 51 cost
Expanded 576 nodes
Max Frontier size 22
Max node memory size 581
Max depth 51
------------------------


In [37]:
!python main.py --maze_id=5 --search_algorithm_name="gbfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze5.txt is loaded
gbfs is loaded
------------------------
Found Path with 51 cost
Expanded 156 nodes
Max Frontier size 52
Max node memory size 207
Max depth 50
------------------------


In [38]:
!python main.py --maze_id=5 --search_algorithm_name="astar"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.7)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze5.txt is loaded
astar is loaded
------------------------
Found Path with 51 cost
Expanded 412 nodes
Max Frontier size 43
Max node memory size 448
Max depth 50
------------------------



__Maze #5__

| algorithm | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| BFS       |   51      |     576        |      51        |      581      |      22           |
| DFS       |   155     |     353        |     198        |      465      |     149           |
| UCS       |   51      |     576        |      51        |      581      |      22           |
| GBFS      |   51      |     156        |      50        |      207      |      52           |
| A*        |   51      |     412        |      50        |      448      |      43           |


Discuss the most important lessons you have learned from implementing these different search strategies. 

#### Discussion

This lab provided a concrete and instructive comparision of uninformed and informed search stategies through their application to different mazes. For maze 5, although several algorithms ultimately found the same optimal solution, their performance in terms of expanded nodes, memmory usage and search behavior differed significantly. While maze 5 provides a clear snapshot of strengths and weakneses, the result from maze 1-4 shows that these behaviors are not incidental but strutural, driven by how each algorithm explores the search space. These differences show the fundamental trade-offs between optimality, efficiency and resource consumption in search algorithms.

Breadth-First Search (BFS) demonstrated highly consistent behavior across all mazes. In every case, BFS found an optimal path, as evidence by the close correspondence between path cost and maximum tree depth. Uniform-Cost Search (UCS) behaved almost identically to BFS across all mazes. BFS and UCS both produced an optimal path with a cost of 51. This outcome is expected because all step costs in the maze are uniform, making UCS functionally equivalent to BFS in this environment. However, both algorithms expanded a large number of nodes (576) and required substantial memory to store the frontier and expanded sets. Due to them being uninformed searches, they expand nodes uniformly in all directions without regard to the heuristic or distance to the goal, which leads to unnecessary computation in large or open ares of the maze. While BFS and UCS achieve optimality when actions cost are all the same, this comes at a high computational cost.

Depth-First Search (DFS) shows the most volatile performance across the mazes. In some cases, such as maze 3, DFS managed to find an optimal path, but this outcome was largely coincidental rather than systematic. In most mazes, DFS produced paths with significantly higher costs and extremely large maximum depths, sometimes exceeding 4 times the optimal solution depth. DFS pursues a single path as deeply as possible before backtracking, which makes it vunerable to getting trapped in long corrdidors or dead ends. This behavior is especially visible in mazes where the goal is not located near the deepest nodes of the search tree, such as maze 1,2,4 and 5, which causes DFS to spend a large amount of time exploring deep, irrelevant branches before eventually encountering the goal. DFS performed the worst overall in maze 5. Although it expanded fewer nodes than BFS and UCS, it produces a significantly longer path with a cost of 155 and reached a maximum depth of 198. DFS returns the first goal found, regardless of depth or cost. Therefore, the resulting solution was highly inefficient, demonstrating that minimizing the number of expanded nodes does not necessarily correlate with finding a good solution. DFS's lack of optimality make it unsuitable for problems where solution quality is important.

Greedy Best-First Search (GBFS) consistently demonstrated the lowest number of expanded nodes and the smallest memory across all mazes. By relying solely on the heuristic being the Manhattan distance to guide the search toward the goal. Inseveral mazes, such maze 3 and maze 5, GBFS even managed to find optimal solutions with dramatically fewer expansions. In maze 5, It found the optimal path with a cost of 51 while expanding only 156 nodes and using the least amount of memory among all algorithms.  The strong performance of GBFS in maze 5 suggests that the heuristic aligned well with the maze's structure, allowing the algorithm to move efficiently toward the target. However, this success should be interpreted cautiously. In mazes 1,2, and 3, GBFS produced paths with higher costs than the optimal solutions. This is a limitation of GBFS. Because it ignores the accumulated path costs, it can be misled by heuristics into choosing locally attractive but globally supoptimal routes. GBFS does not garantee optimal solutions in general and can perform poorly in environments where the heuristic is misleading or where significant detours are required.

A* search gave a balanced compromise between efficiency and reliability across all mazes. In every case, A* found an optimal path while expanding significantly fewer nodes than BFS and UCS. Although it did not match GBFS’s efficiency in terms of node expansion, it consistently required less memory and avoided the severe inefficiencies seen in DFS.Like BFS and UCS in maze 5, A* found the optimal path with a cost of 51, but it expanded significantly fewer nodes (412) and required less memory. By conbining the actual path cost with the heuristic estimate, A* was able to focus the search algorithm toward promising regions while still maintaining optimality garantees. Although it was not as fast as GBFS in this maze, A* provides more consistent performance across different problem structures, making it a robust and generally preferable choice for pathfinding tasks.

Overall, this lab demonstrates that no single search algorithm is universally optimal. Uninformed methods such as BFS and UC provide strong guarantees but suffer from high time and memory costs. DFS is fast in terms of node expansion but unreliable and often suboptimal. Heuristic-based methods darmatically reduce search effort, with GBFS offering impressive efficiency when the heuristic is well suitéd to the problem, and A* providing the best balance between optimality and efficiency. Maze 5 clearly illustrates how heuristic quality and maze structure play a crucial role in determining search performance and emphasizes the importance of selecting an appropriate algorithm based on problem contraints and desired guarantees.

### Task 7: Advanced task only for PhD Students
This task is for **PhD Students only**. Undergraduate students do not need to complete this task.

Modify your A* search to add weights (see textbook and lecture slides) and explore how different weights influence the result. Remember to create a new Python file under ./searchalgorithms.